parlai/core/teachers.py

```diff
@@ -1944,6 +1945,223 @@ def shutdown(self):
             t.shutdown()
 
 
+class ChunkTeacher(FixedDialogTeacher, ABC):
+    """
+    Useful for loading large amounts of data.
+
+    Data is separated into chunks and loaded one chunk at a
+    time. Loads the data off of the main thread.
+    """
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+
+        # make logging print to stdout
+        logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+
+        self.buffersize = 100000
+
+        if 'stream' not in opt['datatype']:
+            raise ValueError(
+                'Chunk teacher should be used with streaming. '
+            )
+        if opt.get('pytorch_teacher_task') is not None:
+            raise ValueError(
+                'Chunk teacher is not compatible with pytorch '
+                'data teacher.'
+            )
+        if opt['numthreads'] > 1:
+            raise ValueError(
+                'Chunk teacher is not compatible with Hogwild.'
+            )
+
+        self.set_datasettings(opt['datatype'])
+        chunks = self.fold_chunks[:]
+
+        if (
+            shared is None
+            and self.is_train
+            and self.opt.get('distributed_world_size') is not None
+        ):
+            dws = int(self.opt['distributed_world_size'])
+            rank = int(self.opt['rank'])
+            self.fold_chunks = [c for c in self.fold_chunks if c % dws == rank]
+
+        if self.is_train:
+            random.Random().shuffle(chunks)
+
+        if shared is not None:
+            self.chunks = shared['chunks']
+            self.samples = shared['samples']
+            self.rng = shared['rng']
+            self.is_root_teacher = False
+        else:
+            self.is_root_teacher = True
+            self.samples = queue.Queue(maxsize=self.buffersize)
+            self.chunks = queue.Queue()
+            if self.is_train:
+                self.rng = random.Random()
+            else:
+                self.rng = random.Random(4)
+            self._enqueue_chunks()
+            # launch queue loader on the main thread
+            self._enqueue_request()
+
+        self.episode_done = True
+
+    def _get_data_folder(self):
+        if not self.opt.get('datafile'):
+            raise RuntimeError(
+                'Must specify datafile or override this function '
+                'to return the data folder.'
+            )
+
+        return self.opt['datafile']
+
+    @abstractmethod
+    def _get_num_samples(self, datatype) -> int:
```

Comment:
abstract methods shouldn't be private. The things I need to implement should be visible to me.

Reply:
But they can still be private and visible to the concrete subclass, which means they'd still warrant an underscore

Reply:
Sure they’re protected. But why are we going to hide the things the user needs to implement from the docs?

Reply:
Hmm - will the docs (for instance, the docs you got to be auto-created Friday) only give documentation for public methods? In that case, yeah, maybe it makes sense to make this method public to increase its visibility